File: android/app/src/main/kotlin/com/aibook/android/feature/opds/OpdsSyncCollector.kt
package com.aibook.android.feature.opds

import com.aibook.android.core.network.opds.OpdsEntry
import com.aibook.android.core.network.opds.OpdsFeed

class OpdsSyncCollector(
    private val maxDepth: Int = DEFAULT_MAX_DEPTH,
    private val maxPagesPerCatalog: Int = DEFAULT_MAX_PAGES_PER_CATALOG,
    private val loadFeed: suspend (href: String?) -> OpdsFeed
) {
    suspend fun collect(): OpdsSyncCollection {
        val visitedPages = mutableSetOf<String>()
        val acquisitionEntries = linkedMapOf<String, OpdsEntry>()
        var catalogCount = 0

        suspend fun visit(href: String?, depth: Int) {
            if (depth > maxDepth) return
            var pageHref = href
            var pageCount = 0
            while (pageCount < maxPagesPerCatalog) {
                val pageKey = pageHref ?: "__root__"
                if (!visitedPages.add(pageKey)) break
                val feed = loadFeed(pageHref)
                feed.entries.filter { it.acquisitionLink != null }.forEach { entry ->
                    val key = entry.identifier ?: entry.acquisitionLink?.href ?: return@forEach
                    acquisitionEntries[key] = entry
                }
                for (entry in feed.entries.filter { it.acquisitionLink == null }) {
                    val childHref = entry.alternateLink?.href?.takeIf { it.isNotBlank() } ?: continue
                    if (childHref in visitedPages) continue
                    catalogCount += 1
                    visit(childHref, depth + 1)
                }
                pageHref = feed.nextLink?.href?.takeIf { it.isNotBlank() } ?: break
                pageCount += 1
            }
        }

        visit(href = null, depth = 0)
        return OpdsSyncCollection(
            acquisitionEntries = acquisitionEntries.values.toList(),
            catalogCount = catalogCount
        )
    }

    companion object {
        const val DEFAULT_MAX_DEPTH = 4
        const val DEFAULT_MAX_PAGES_PER_CATALOG = 100
    }
}

data class OpdsSyncCollection(
    val acquisitionEntries: List<OpdsEntry>,
    val catalogCount: Int
)

File: ios/AiBook/Core/Network/Opds/OpdsSyncCollector.swift
import Foundation

// MARK: - OpdsSyncCollector（与安卓 OpdsSyncCollector 对齐 — 递归遍历 OPDS 目录树）

final class OpdsSyncCollector {
    private let catalogService: OpdsCatalogService
    private let maxDepth: Int

    init(catalogService: OpdsCatalogService = OpdsCatalogService(), maxDepth: Int = 4) {
        self.catalogService = catalogService
        self.maxDepth = maxDepth
    }

    struct SyncCollection {
        let entries: [OpdsEntry]
        let totalCount: Int
    }

    /// 递归收集所有可获取的书籍条目
    func collect(
        from url: URL,
        username: String? = nil,
        password: String? = nil,
        depth: Int = 0
    ) async throws -> SyncCollection {
        guard depth < maxDepth else {
            return SyncCollection(entries: [], totalCount: 0)
        }

        let feed = try await catalogService.loadFeed(url: url, username: username, password: password)

        var allEntries: [OpdsEntry] = []

        for entry in feed.entries {
            if entry.acquisitionLink != nil {
                // 这是一个可下载的书籍
                allEntries.append(entry)
            } else if let alternateHref = entry.alternateLink?.href,
                      let alternateUrl = URL(string: alternateHref, relativeTo: url) {
                // 这是一个子目录，递归遍历
                let subCollection = try await collect(
                    from: alternateUrl,
                    username: username,
                    password: password,
                    depth: depth + 1
                )
                allEntries.append(contentsOf: subCollection.entries)
            }
        }

        return SyncCollection(entries: allEntries, totalCount: allEntries.count)
    }
}
